src/services/pasteEdits.ts

```diff
@@ -104,12 +107,45 @@ function pasteEdits(
                 preferences,
                 formatContext,
             };
-            forEachChild(updatedFile, function cb(node) {
-                if (isIdentifier(node) && !originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
-                    // generate imports
-                    importAdder.addImportForUnresolvedIdentifier(context, node, /*useAutoImportProvider*/ true);
-                }
-                node.forEachChild(cb);
+
+            // `updatedRanges` represent the new ranges that account for the offset changes caused by pasting new text and
+            // `offset` represents by how much the starting position of `pasteLocations` needs to be changed.
+            //
+            // We iterate over each updated range to get the node that wholly encloses the updated range. For each child of that node, it is checked if the
+            // identifier lies within the updated range and if it is not resolved, we try resolving it.
+            let offset = 0;
+            pasteLocations.forEach((location, i) => {
+                const oldTextLength = location.end - location.pos;
+                const textToBePasted = actualPastedText ? actualPastedText[0] : pastedText[i];
+                const startPos = location.pos + offset;
+                const endPos = startPos + textToBePasted.length;
+                const range = { pos: startPos, end: endPos } satisfies TextRange;
+                offset += textToBePasted.length - oldTextLength;
+                const enclosingNode = findAncestor(
+                    getTokenAtPosition(context.sourceFile, range.pos),
+                    ancestorNode => rangeContainsRange(ancestorNode, range),
+                );
+                if (!enclosingNode) return;
+                forEachChild(enclosingNode, function importUnresolvedIdentifiers(node) {
+                    if (
+                        isIdentifier(node) && rangeContainsPosition(
+                            range,
+                            node.getStart(updatedFile),

```

Comment:
Do you need to run `node.getStart()`? Can you just use `node.pos`?

Reply:
Yes, `node.pos` gives the position with the whitespaces. So, for example in pasteEdits_multiplePastesConsistentlyLargerInSize.ts for the 3rd paste location (`const test = [|1 + 2|] + 3;` ) , `node.pos` will give the position of the whitespace before 1, but the updated range of the pasted text starts from 1. As `node.pos` will not lie in this range, it will skip checking for the pasted text here. `node.Start()` avoids this by giving the exact position of 1